src/trustRegion.jl

```diff
@@ -0,0 +1,376 @@
+    if max_trust_radius == 0.0
+        max_trust_radius = max(norm(fu), maximum(u) - minimum(u))
+    end
+    if initial_trust_radius == 0.0
+        initial_trust_radius = max_trust_radius / 11
+    end
+
+    return TrustRegionCache{iip}(f, alg, u, fu, p, uf, linsolve, J, du1, jac_config,
+                                 1, false, maxiters, internalnorm,
+                                 ReturnCode.Default, abstol, prob, initial_trust_radius,
+                                 max_trust_radius, loss, loss, copy(J), fu, 0, u, u, fu, true,
+                                 0.0, u)
+end
+
+function perform_step!(cache::TrustRegionCache{true})
+    @unpack make_new_J, J, fu, f, u, p, du1, alg, linsolve = cache
+    if cache.make_new_J
+        jacobian!(J, cache)
+        mul!(cache.H, J, J)
+        mul!(cache.g, J, fu)
+    end
+
+    # u = u - J \ fu
+    linres = dolinsolve(alg.precs, linsolve, A = cache.H, b = _vec(cache.g),
+                        linu = _vec(du1),
+                        p = p, reltol = cache.abstol)
+    cache.linsolve = linres.cache
+    cache.δN = -du1
+    dogleg!(cache)
+
+    # Compute the potentially new u
+    cache.u_new .= u .+ cache.step_size
+    f(cache.fu_new, cache.u_new, p)
+
+    trust_region_step!(cache)
+    return nothing
+end
+
+function perform_step!(cache::TrustRegionCache{false})
+    @unpack make_new_J, fu, f, u, p = cache
+
+    if make_new_J
+        J = jacobian(cache, f)
+        cache.H = J * J
+        cache.g = J * fu
+    end
+
+    @unpack g, H = cache
+    # Compute the Newton step.
+    cache.δN = -H \ g
+    dogleg!(cache)
+
+    # Compute the potentially new u
+    cache.u_new = u .+ cache.step_size
+    cache.fu_new = f(cache.u_new, p)
+
+    trust_region_step!(cache)
+    return nothing
+end
+
+function trust_region_step!(cache::TrustRegionCache)
+    @unpack fu_new, u_new, step_size, g, H, loss, alg, max_trust_r = cache
+    cache.loss_new = get_loss(fu_new)
+
+    # Compute the ratio of the actual reduction to the predicted reduction.
+    cache.r = -(loss - cache.loss_new) / (step_size' * g + step_size' * H * step_size / 2)
+    @unpack r = cache
+
+    # Update the trust region radius.
+    if r < alg.shrink_threshold
+        cache.trust_r *= alg.shrink_factor
+        cache.shrink_counter += 1
+    else
+        cache.shrink_counter = 0
+    end
+    if r > alg.step_threshold
+
+        # Take the step.
+        cache.u = u_new
+        cache.fu = fu_new
+        cache.loss = cache.loss_new
+
+        # Update the trust region radius.
+        if r > alg.expand_threshold
+            cache.trust_r = min(alg.expand_factor * cache.trust_r, max_trust_r)
+        end
+
+        cache.make_new_J = true
+    else
+        # No need to make a new J, no step was taken, so we try again with a smaller trust_r
+        cache.make_new_J = false
+    end
+
+    if iszero(cache.fu) || cache.internalnorm(cache.step_size) < cache.abstol
+        cache.force_stop = true
+    end
+end
```

Comment:
I don't see `du1` used in here.  So can't we just do `cache.du1 .= -1 .* du1` and then use it in here to eliminate a tmp vector?

Reply:
sure

Reply:
Hmm... Tried to replace $\delta N$, but did you have another vector in mind?

Reply:
I've now replaced $\delta N$, du1 and u_new with a single vector u_tmp